README.md
# Math::GSL

Math::GSL is a Perl interface to the [GNU Scientific Library](http://www.gnu.org/software/gsl/), using [SWIG](http://swig.org).  The GNU
Scientific Library (GSL) is a numerical library for C and C++ programmers. It
is free software under the GNU General Public License.  Math::GSL uses SWIG to
generate Perl bindings to *most* GSL functionality.

[![Build Status](https://secure.travis-ci.org/leto/math--gsl.png)](http://travis-ci.org/leto/math--gsl)

# Dependencies

## CPAN distribution dependencies

Currently Math::GSL requires at least Perl 5.8.1 and GSL 1.15 to compile. It
also requires that the gsl-config binary can be found in your PATH or [PkgConfig](http://metacpan.org/release/PkgConfig/) to find GSL.


## Git repo dependencies

SWIG >= 2.x is needed to build Math::GSL from the git repo, in addition to all
the dependencies above. 2.0.8 or newer is required to work with Perl 5.20 and
higher. SWIG 3.x is recommended.

On OS X with Homebrew, you can install swig with:

    brew install swig

# Installation

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install clean

# Support

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Math::GSL

You can also look for information at:

MetaCPAN: https://metacpan.org/release/Math-GSL

Known bugs/issues: https://github.com/leto/math--gsl/issues/

CPAN Ratings http://cpanratings.perl.org/d/Math::GSL


# Copyright and Licence

Copyright (C) 2008-2018 Jonathan "Duke" Leto and Thierry Moisan.

A full list of contributors is listed in the [CREDITS](https://github.com/leto/math--gsl/blob/master/CREDITS) file.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself. Fuck Yeah.


Changes
=head1 Math::GSL Changes

=head1 v0.41 - xxx 2018
    - Add support for Debian-specific version 1.16.1, which caused test failures

=head1 v0.40 - December 2018
    - added support for GSL 2.5 (djerius)
    - vast improvements to Travis CI test infrastructure (djerius)
    - added random method in Math::GSL::Matrix (Rui Meira)
    - added Math::GSL::Multiset (standard gsl API and Perl OO)
    - gsl_linalg_givens now works
    - gsl_vector_const_subvector is now exported properly
    - Support for Sparse Matrices was added
    - Documentation for Math::GSL::Rstat, Math::GSL::SparseMatrix,
      Math::GSL::Diff
    - Math::GSL::Multifit now requires GSL 2.1 and higher
    - Math::GSL::Rstat now requires GSL 2.1 and higher

=head1 v0.39 - September 2016
    - Fix Debian Bug reported by Niko Tyni
        https://bugs.debian.org/826009
    - Extensive fixes to allow GSL 1.15 and GSL 1.16
      to continue to work with GSL 2.x
    - GSL 1.11 - 1.14 are no longer supported

=head1 v0.38 - September 2016
    - Fixes for compiling against various GSL
      versions (thanks CPANtesters!)

=head1 v0.37 - September 2016
    - Support for GSL 2.0, 2.1 and 2.2
    - Multilarge subsystem added for GSL >= 2.1

=head1 v0.36 - Fri Mar 25 23:32:03 PDT 2016
    - Fixed error messages on Matrix set_col and set_row
      methods (Jon Schutz)
    - Implemented array-based Bessel functions:
               gsl_sf_bessel_Jn_array
               gsl_sf_bessel_Yn_array
               gsl_sf_bessel_In_array
               gsl_sf_bessel_In_scaled_array
               gsl_sf_bessel_Kn_array
               gsl_sf_bessel_Kn_scaled_array
               gsl_sf_bessel_jl_array
               gsl_sf_bessel_jl_steed_array
               gsl_sf_bessel_yl_array
               gsl_sf_bessel_il_scaled_array

=head1 v0.35 - April 2015

    - Remotion of trailing whitespaces (Shlomi Fish)
    - Documentation fixes (Shlomi Fish)
    - Remotion of PkgConfig from runtime requirements
      (Bartosz Jakubski)
    - Indentation fixes (Bartosz Jakubski)
    - Installation feedback improvements (Bartosz Jakubski)

=head1 v0.34 - December 2014

    - Fixed bug on uncleared error variable (5.8.9)
    - Added OO transpose method for matrices (Ivan Baidakou)

=head1 v0.33 - September 2014
    - Added OO code for RNG->shuffle, RNG->choose and RNG->sample.
    - Complain when opening (fopen) a file from a tainted string.
    - Complex now has operator overloading for +,-, ==, !=, *, /
    - Added write/load methods to save a matrix to file.

=head1 v0.32 - June 2014
    - Correctly split the compiler name.
    - Added OO code to the QRNG module.
    - Fixed QRNG code that was failing on some machines.

=head1 v0.31 - June 2014
    - Fixed Matrix multiplication that was forcing square matrices.

=head1 v0.30 - June 2014
    - Fixed generated C files with wrong GSL version numbers.
    - Refactored Build.PL in order to use either gsl-config or PkgConfig.
    - Change Build.PL to return 0 when GSL library not found.
    - Use PkgConfig instead of ExtUtils::PkgConfig.
    - Math::GSL::Matrix constructor from a GSL matrix.
    - Reorganization on the Math::GSL::Matrix documentation.
	- Add symbols introduced on gsl 1.15.
	- Fixed in the correct way the prove/TAP::Harness problem.

=head1 v0.29 - June 2014
	- Fix issue when prove/TAP::Harness uses lib/ modules instead of blib/ ones.
    - Add functions introduced in gsl 1.16.
    - Fix issue when compiling on some Linuxes.
	- Math::GSL ->each method.
    - Math::GSL ->max and ->min methods.
	- Math::GSL ->dim method.
    - Math::GSL ->set_elem and ->get_elem methods.
    - Math::GSL concat_vertically and concat_horizontally.

=head1 v0.28 - Jan 2014
    - Fix build errors with clang
    - Improve POD to play nice with Perl 5.18
    - Teach Math::GSL about new functions added in GSL 1.16
    - Added 1D minimization tests

=head1 v0.27 - July 2012
    - Mathieu functions are now accessible
    - Improved detection of gsl
    - Fix bug in calling ->get() method on Math::GSL::Vectors
    - Greatly improved HTML formatting of documentation

=head1 v0.26 - September 2011
    - The Math::GSL::Min and Math::GSL::Roots are now functional - Yury Zavarin
    - Dirichlet random distribution support and greatly increased docs by Tom Nishimura <tnish@fastmail.jp>
    - gsl_version() function which returns the version of GSL being used by Tom Nishimura <tnish@fastmail.jp>
    - Fix '-bundle not allowed with -dynamiclib' build error on Mac OS X - Yury Zavarin
    - Build swig wrappers for different versions of GSL (1.11 - 1.15) - Yury Zavarin
    - Minimum SWIG version is now 2.x (only required for developers using the git repo)
    - Removed deprecated stuff from Build.PL and added MyModule::Build to configure_requires
    - Many small documentation improvements
    - Fix build on Windows - Sisyphus

=head1 v0.25 - April 2011
    - TODO RT#66882 properly so that a stable release with a passing test suite can be used.

=head1 v0.24 - April 2011
    - Fix https://rt.cpan.org/Ticket/Display.html?id=66882 (Tests fail on Perl's with DUSELONGDOUBLE)

=head1 v0.23 - March 2011
    - Fixed the tolerance of some tests that were causing CPANtesters failures
    - Fixed some doc typos

=head1 v0.22 - June 2010
    - Improved tests
    - == and != operators for Vectors
    - Fixed the export of hypergeometric functions (Thierry Moisan)
    - In instantiation of GSLBuilder in Build.PL, the value for the parameter
      include_dirs were corrected to an anonymous empty array, instead of an
      empty string.  (Jakob Ilves)
    - Operator overloading of vector equality (Thierry Moisan)
    - Various doc fixes (Pjotr Prins)
    - Typemap fixes (Vincent Danjean)

=head1 v0.20 - May 19 2009
    - $rng->get($n) will now return the next $n values of the random number generator
    - Fixed + added test for RT#45044 math-gsl eigenpair bug, reported with patch by Ian Malone
    - Added more tests for Randist

.build_gsl.sh
#! /bin/bash

get_gsl_version() {
   cd $GSL_SRC_DIR

    echo "Retrieving GSL $1"
    wget -q $GSL_MIRROR/gsl-$1.tar.gz \
         --retry-connrefused \
         --timeout=900 

    if [[ -d gsl-$1 ]] ; then
        rm -rf gsl-$1
    fi
    echo
    echo "Extracting..."
    tar zxpf gsl-$1.tar.gz
    cd gsl-$1
    echo
    echo "Configuring..."
    ./configure --prefix $(readlink -f $GSL_INST_DIR/gsl-$1) 
    echo
    echo "Building..."
    make -j2 >& make.log
    echo
    echo "Installing..."
    make -j2 install >& install.log
    rm -rf gsl-$1
}

export GSL=2.5
export GSL_CURRENT=2.5
export GSL_INST_DIR=$(readlink -f gsl)
export GSL_SRC_DIR="$GSL_INST_DIR/src"
export DIST_DIR=$GSL_INST_DIR
export GSL_MIRROR=http://mirrors.kernel.org/gnu/gsl
mkdir -p $GSL_SRC_DIR

get_gsl_version $GSL


inc/GSLBuilder.pm
package GSLBuilder;

use strict;
use warnings;
use Config;
use File::Copy;
use File::Path qw/mkpath/;
use File::Spec::Functions qw/:ALL/;
use base 'MyModule::Build';

use Ver2Func;

sub is_release {
    return -e '.git' ? 0 : 1;
}

sub process_swig_files {
    my $self = shift;
    my $p = $self->{properties};

    unless (is_release()) {
        $self->process_versioned_swig_files;
    }

    my $binding_ver     = $self->get_binding_version;
    my $current_version = $self->{properties}->{current_version};
    my $gsl_prefix      = $self->{properties}->{gsl_prefix};
    my $gsl_ldflags     = $self->{properties}->{extra_linker_flags};
    my $gsl_ccflags     = $self->{properties}->{extra_compiler_flags};
    my $swig_flags      = $self->{properties}->{swig_flags};
    my $swig_version    = $self->{properties}->{swig_version};

    if ($binding_ver ne $current_version) {
        print "VERSION MISMATCH: Let's hope for the best.\n";
    }
    print "Processing $binding_ver XS files, GSL $current_version (via gsl-config) at $gsl_prefix\n";
    print "Compiler        = " . qx{ $Config{cc} --version } . "\n";
    print "ccflags         = @$gsl_ccflags\n";
    print "ldflags         = @$gsl_ldflags\n";
    print "swig_flags      = $swig_flags\n" unless is_release();
    print "swig_version    = $swig_version\n" unless is_release();
    my $PERL5LIB           = $ENV{PERL5LIB} || "";
    my $LD_LIBRARY_PATH    = $ENV{LD_LIBRARY_PATH} || "";
    print "PERL5LIB        = $PERL5LIB\n";
    print "LD_LIBRARY_PATH = $LD_LIBRARY_PATH\n";

    $self->process_xs_file( $_, $binding_ver )
      foreach Ver2Func->new( $current_version )->swig_files;
}

sub get_binding_version {
    my $self = shift;
    my $current_version = $self->{properties}->{current_version};
    my @bindings = <xs/*>;
    my $all_binding_versions = {};
    foreach my $b (@bindings) {
        if ($b =~ /\w(?:\d+)?\.([\d\.]+)\.c$/) {
            $all_binding_versions->{$1} = 1;
        }
    }
    my $result_binding_version;

    # make sure we find 2.2.1 before 2.2
    foreach my $b_ver (reverse sort keys %{$all_binding_versions}) {
        if (cmp_versions($current_version, $b_ver) >= 0 &&
            (!defined($result_binding_version) ||
             cmp_versions($result_binding_version, $b_ver) == -1))
        {
            $result_binding_version = $b_ver;
        }
    }

    unless (defined($result_binding_version)) {
        die "Can't find appropriate bindings version, ".
            "check 'xs' directory, it should contains bindings " .
            "with version <= current GSL version ($current_version)";
    }

    return $result_binding_version;
}

sub process_versioned_swig_files {
    my $self = shift;

    my $p = $self->{properties};

    my $cur_ver = $p->{current_version};

    foreach my $ver ( Ver2Func->versions( $cur_ver ) ) {

        print "Building wrappers for GSL $ver\n";

        my $ver2func = Ver2Func->new( $ver );

	my $renames_i = catfile( qw/swig renames.i/ );
        $ver2func->write_renames_i( $renames_i );
	copy( $renames_i, catfile( 'swig', "renames.${ver}.i" ) );

        foreach my $entry ( $ver2func->sources ) {
            my ( $swig_file, $deps ) = @$entry;
            $self->process_swig( $swig_file, $deps, $ver );
        }
    }
}

sub process_xs_file {
    my ($self, $main_swig_file, $ver) = @_;

    (my $file_base = $main_swig_file) =~ s/\.[^.]+$//;
    $file_base =~ s!\\!/!g;
    (undef, undef, $file_base) = splitpath($file_base);

    my $c_file = catfile('xs',"${file_base}_wrap.$ver.c");

    # .c -> .o
    my $obj_file = $self->compile_c($c_file);
    $self->add_to_cleanup($obj_file);

    my $archdir = catdir($self->blib, qw/arch auto Math GSL/, $file_base);
    mkpath $archdir unless -d $archdir;

    # .o -> .so
    $self->link_c($archdir, $file_base, $obj_file);

    my $from = catfile(qw/pm Math GSL/, "${file_base}.pm.$ver");
    my $to = catfile(qw/blib lib Math GSL/, "${file_base}.pm");
    chmod 0644, $from, $to;
    copy($from, $to);
}

sub cmp_versions {
    my ($v1, $v2) = @_;
    my @v1 = split(/\./, $v1);
    my @v2 = split(/\./, $v2);
    my $cmp_major = $v1[0] <=> $v2[0];
    return $cmp_major if ($cmp_major != 0);
    my $cmp_minor = $v1[1] <=> $v2[1];
    return $cmp_minor;
}

# Check dependencies for $main_swig_file. These are the
# %includes. If needed, arrange to run swig on $main_swig_file to
# produce a xxx_wrap.c C file.
sub process_swig {
    my ($self, $main_swig_file, $deps_ref, $ver) = @_;
    my ($cf, $p) = ($self->{config}, $self->{properties});

    (my $file_base = $main_swig_file) =~ s/\.[^.]+$//;
    $file_base =~ s!swig/!!g;
    my $c_file = catfile('xs',"${file_base}_wrap.$ver.c");

    my @deps = defined $deps_ref ?  @$deps_ref : ();

    # don't bother with swig if this is a CPAN release
    $self->compile_swig($main_swig_file, $c_file, $ver)
        unless $self->up_to_date([$main_swig_file, @deps], $c_file);
}

sub swig_binary_name {
    # recent versions of Ubuntu call it swig2.0 . Le sigh.
    my $cmd = "swig -version";
    my $out  = eval { no warnings; `$cmd` };
    if ($?) {
        $cmd = "swig2.0 -version";
        $out  = eval { no warnings; `$cmd` };
        if ($?) {
            $cmd = "swig3.0 -version";
            $out = eval { no warnings; `$cmd` };

            if ($?) {
                die "Can't find the swig binary!";
            } else {
                return "swig3.0";
            }
        }
        return "swig2.0";
    }
    return "swig";
}

# Invoke swig with -perl -outdir and other options.
sub compile_swig {
    my ($self, $file, $c_file, $ver) = @_;
    my ($cf, $p) = ($self->{config}, $self->{properties}); # For convenience

    # File name, minus the suffix
    (my $file_base = $file) =~ s/\.[^.]+$//;

    # get rid of the swig name
    $file_base =~ s!swig/!!g;

    my $pm_file = "${file_base}.pm";

    my @swig       = swig_binary_name(), defined($p->{swig}) ? ($self->split_like_shell($p->{swig})) : ();
    my @swig_flags = defined($p->{swig_flags}) ? $self->split_like_shell($p->{swig_flags}) : ();

    my $blib_lib = catdir(qw/blib lib/);
    my $gsldir = catdir('pm', qw/Math GSL/);
    mkpath $gsldir unless -e $gsldir;

    my $from    = catfile($gsldir, $pm_file);
    my $to      = catfile(qw/lib Math GSL/, $pm_file);
    chmod 0644, $from, $to;

    my @args = ( @swig, '-o', $c_file , '-outdir', $gsldir, '-perl5', @swig_flags, $file);
    print join(" ", @args ) . "\n";
    $self->do_system( @args ) or die "error : $! while building ( @swig_flags ) $c_file in $gsldir from '$file'";
    move($from, "$from.$ver");

    {
      ## updates the version number.
      ## all files are being processed right now.
      ## later versions might use a fixed list of candidate files.
      undef $/;
      open my $in, "<", $c_file or die "Can't open file $c_file: $!";
      my $contents = <$in>;
      close $in;

      # TODO: this doesn't support x.y.z
      $contents =~ s{("GSL_VERSION", TRUE \| 0x2 \| GV_ADDMULTI\);[\s\n]*sv_setsv\(sv, SWIG_FromCharPtr\(")\d\.\d+}
                    {$1$ver};

      open my $out, ">", $c_file or die "Can't overwrite file $c_file: $!";
      print $out $contents;
      close $out;
    }

    if ($p->{current_version} eq $ver) {
        # print "Copying from: $from.$ver, to: $to; it makes the CPAN indexer happy.\n";
        copy("$from.$ver", $to);
    }

   return $c_file;
}
sub is_windows { $^O =~ /MSWin32/i }
sub is_darwin  { $^O =~ /darwin/i  }
sub is_cygwin  { $^O =~ /cygwin/i }

# Windows fixes courtesy of <sisyphus@cpan.org>
sub link_c {
  my ($self, $to, $file_base, $obj_file) = @_;
  my ($cf, $p) = ($self->{config}, $self->{properties}); # For convenience

  my $lib_file = catfile($to, File::Basename::basename("$file_base.$Config{dlext}"));
  # this is so Perl can look for things in the standard directories
  $lib_file =~ s!swig/!!g;

  $self->add_to_cleanup($lib_file);
  my $objects = $p->{objects} || [];

  unless ($self->up_to_date([$obj_file, @$objects], $lib_file)) {
    my @linker_flags = $self->split_like_shell($p->{extra_linker_flags});
    push @linker_flags, "-v" if $ENV{DEBUG_LD};

    if(is_windows() or is_darwin()) {
      if(is_windows() && $] eq '5.010000' && $Config{archname} =~ /x64/) {
         push @linker_flags, $Config{bin} . '/' . $Config{libperl};
      }
      else {
        push @linker_flags, $Config{archlib} . '/CORE/' . $Config{libperl};
      }
    }

    my @lddlflags = $self->split_like_shell($cf->{lddlflags});
    my @shrp      = $self->split_like_shell($cf->{shrpenv});
    my @ld        = $self->split_like_shell($cf->{ld} || $Config{cc});

    # Strip binaries if we are compiling on windows
    push @ld, "-s" if (is_windows() && $Config{cc} =~ /\bgcc\b/i);

    print join(" ", @shrp, @ld, @lddlflags, '-o', $lib_file, $obj_file, @$objects, @linker_flags) . "\n";
    $self->do_system(@shrp, @ld, @lddlflags, '-o', $lib_file,
		     $obj_file, @$objects, @linker_flags)
      or die "error building $lib_file file from '$obj_file'";
  }

  return $lib_file;
}

# From Base.pm but modified to put package cflags *after*
# installed c flags so warning-removal will have an effect.

sub compile_c {
  my ($self, $file) = @_;
  my ($cf, $p) = ($self->{config}, $self->{properties}); # For convenience

  # File name, minus the suffix
  (my $file_base = $file) =~ s/\.[^.]+$//;
  my $obj_file = $file_base . $Config{_o};

  $self->add_to_cleanup($obj_file);
  return $obj_file if $self->up_to_date($file, $obj_file);

  $cf->{installarchlib} = $Config{archlib};

  my @include_dirs = @{$p->{include_dirs}}
			? map {"-I$_"} (@{$p->{include_dirs}}, catdir($cf->{installarchlib}, 'CORE'))
			: map {"-I$_"} ( catdir($cf->{installarchlib}, 'CORE') ) ;

  my @extra_compiler_flags = $self->split_like_shell($p->{extra_compiler_flags});
  my @cccdlflags           = $self->split_like_shell($cf->{cccdlflags});
  my @ccflags              = $self->split_like_shell($cf->{ccflags});

  push @ccflags, $self->split_like_shell($Config{cppflags});
  my @optimize = $self->split_like_shell($cf->{optimize});

  # Whoah! There seems to be a bug in gcc 4.1.0 and optimization
  # and swig. I'm not sure who's at fault. But for now the simplest
  # thing is to turn off all optimization. For the kinds of things that
  # SWIG does - do conversions between parameters and transfers calls
  # I doubt optimization makes much of a difference. But if it does,
  # it can be added back via @extra_compiler_flags.

  my @flags = (@include_dirs, @cccdlflags, '-c', @ccflags, @extra_compiler_flags, );

  my @cc = $self->split_like_shell($cf->{cc});
  @cc = $self->split_like_shell($Config{cc}) unless @cc;
  print join(" ", @cc, @flags, '-o', $obj_file, $file) . "\n";
  $self->do_system(@cc, @flags, '-o', $obj_file, $file)
    or die "error building $Config{_o} file from '$file'";

  return $obj_file;
}

# Propagate version numbers to all modules
sub get_metadata {
    my ($self, @args) = @_;
    my $data = $self->SUPER::get_metadata(@args);
    for my $mod (values %{$data->{provides}}) {
        $mod->{version} ||= 0;
    }
    return $data;
}

3.14;


Build.PL
#!/usr/bin/perl -w
# This has been heavily modified from the Device::Cdio Build.PL
# Jonathan "Duke" Leto <jonathan@leto.net>

#    Copyright (C) 2006 Rocky Bernstein <rocky@cpan.org>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
use strict;
use warnings;
use Config;
use Data::Dumper;
use MyModule::Build;
use lib 'inc';
use GSLBuilder;
use Ver2Func;
use File::Spec::Functions qw/:ALL/;

our $MIN_GSL_VERSION = "1.15";

my $gsl_conf;

$gsl_conf = check_gsl_version();
$gsl_conf = check_gsl_version_pkgconfig() if not $gsl_conf;

my ($major,$minor,$tiny) = split /\./, $gsl_conf->{gsl_version};
my $ccflags = $gsl_conf->{gsl_cflags};
my $ldflags = $gsl_conf->{gsl_libs};

# In case GSL in installed in the system-wide directory, $ccflags is
# empty (because pkg-config remove -I/usr/include), but swig needs it

my $swig_flags = '';
$swig_flags = ' -I/usr/include' if $ccflags =~ /^\s*$/;

$ccflags .= " $Config{ccflags}"; # we need this for some Linuxes

## Swig produces a number of GCC warnings. Turn them off if we can.
$ccflags .= try_cflags("-Wall");
$ccflags .= try_cflags("-Wno-sometimes-uninitialized");
#$ccflags .= try_cflags("-Wno-strict-aliasing");
$ccflags .= try_cflags("-Wno-unused-function");
$ccflags .= try_cflags("-Wno-unused-value");
$ccflags .= try_cflags("-Wno-unused-function");
$ccflags .= try_cflags("-Wno-unused-variable");
$ccflags .= try_cflags("-Wno-gnu");
$ccflags .= try_cflags("-g");

$swig_flags = "$gsl_conf->{gsl_cflags} $swig_flags";

if ( GSLBuilder::is_cygwin()  && $Config{shrpenv} =~ m{\Aenv LD_RUN_PATH=(.*)\Z} ) {
    $ldflags .= " -L$1 -lperl";
    # Should we check the 32-ness?
    $swig_flags = '-DNEED_LONG';
} elsif (GSLBuilder::is_darwin()) {
    $ldflags .= ' -bundle -flat_namespace ';
}

if ($Config{archname} =~ /x86_64|amd64/ ) {
    $ldflags .= ' -fPIC  -fno-omit-frame-pointer ';
    $ccflags .= ' -fPIC  -fno-omit-frame-pointer ';
}
# this causes warnings with clang + OSX
$ccflags = $Config{cccdlflags} . ' ' . $ccflags unless GSLBuilder::is_darwin();

my $ver2func = Ver2Func->new( $gsl_conf->{gsl_version} );

my $cleanup = qq{
                 xs/*_wrap*.c core *.core swig/system.i swig/renames.i
                 swig/*.o Makefile Math-GSL-* tmp* pod2ht*.tmp _build pm
                 lib/Math/GSL/[A-z]+/* blib *.so *.orig
                } . join " ", map { catfile( (qw/lib Math GSL/, "$_.pm") ) } $ver2func->subsystems;


$ldflags = '-shared ' . $ldflags unless GSLBuilder::is_darwin();

my $swig_version;

unless (GSLBuilder::is_release()) {
    $swig_version = swig_version();
    if ((split(/\./, $swig_version))[0] < 2) {
        print "
***
*** You need to have SWIG 2.0 or greater installed. (You have $swig_version).
*** Get SWIG at http://www.swig.org/\n";
        exit 0;
    }
}

$ccflags = "$ccflags " . ($ENV{CC_FLAGS}||''),

my $builder = GSLBuilder->new(
    module_name         => 'Math::GSL',
    add_to_cleanup      => [ $cleanup ],
    dist_abstract       => 'Interface to the GNU Scientific Library using SWIG',
    dist_author         => 'Jonathan Leto <jonathan@leto.net>',
    dist_version_from   => catfile(qw/lib Math GSL.pm/),
    include_dirs        => [],
    extra_linker_flags  => $ldflags,
    extra_compiler_flags=> $ccflags . " " . $swig_flags,
    # suppress swig warnings about GSL source code
    swig_flags          => $swig_flags . " -w451,454",
    license             => 'gpl',
    meta_merge     => {
        resources => {
            repository  => 'http://github.com/leto/math--gsl/tree/master'
        },
    },
    build_requires => {
        'Test::Most'      => '0.31',
        'Test::More'      => 0,
        'Test::Exception' => '0.21',
        'Test::Class'     => '0.12',
        'Test::Taint'     => '1.06',
    },
    requires => {
        'Math::Complex'   => 0,
        'Scalar::Util'    => 0,
        'version'         => '0.77',
        'perl'            => '5.8.1',
        'parent'          => 0,
    },
    sign                  => 0,
    configure_requires => {
        'PkgConfig'       => '0.07720',
        'MyModule::Build'   => '0.38',
        'Alien::GSL'      => '1.01',
    },
    swig_version          => $swig_version,
    current_version       => $gsl_conf->{gsl_version},
    gsl_prefix            => $gsl_conf->{gsl_prefix},
);


$builder->add_build_element('swig');

$builder->create_build_script();

print "Have a great day!\n";


### aux functions

sub swig_version {
    my $swig = GSLBuilder::swig_binary_name();
    my $cmd  = "$swig -version";
    my $out  = `$cmd`;

    if ($?) {
        die "Can't run [$cmd] : $out";
    }
    if ($out =~ /swig\s+version\s+([\d\.]+)\n/i) {
        return $1;
    } else {
        die "Can't parse SWIG version from this output:\n$out";
    }
}

sub modify_env {
    $ENV{PKG_CONFIG_PATH} ||= '';

    # TODO: What about windows and darwin?
    my @guess = qw(/usr/pkgconfig /usr/lib/pkgconfig /usr/local/lib/pkgconfig
                   /usr/local/pkgconfig /usr/libdata/pkgconfig
                   /usr/local/libdata/pkgconfig /opt/pkgconfig);

    unshift @guess, $ENV{PKG_CONFIG_PATH} if $ENV{PKG_CONFIG_PATH};

    $ENV{PKG_CONFIG_PATH} = join(":", @guess);
}

sub try_compile {
    my ($c, %args) = @_;

    my $ok = 0;
    my $tmp = "tmp$$";
    local(*TMPC);

    my $obj_ext = $Config{_o} || ".o";
    unlink("$tmp.c", "$tmp$obj_ext");

    if (open(TMPC, ">", "$tmp.c")) {
    print TMPC $c;
    close(TMPC);

    my $cccmd = $args{cccmd};
    my $errornull;
    my $ccflags =  $Config{'ccflags'};
    $ccflags .= " $args{ccflags}" if $args{ccflags};

    if ($args{silent} ) {
        $errornull = "2>/dev/null" unless defined $errornull;
    } else {
        $errornull = '';
    }

        $cccmd = "$Config{'cc'} -o $tmp $ccflags $tmp.c $errornull"
        unless defined $cccmd;

    printf "cccmd = $cccmd\n" if $args{verbose};
    my $res = system($cccmd);
    $ok = defined($res) && $res == 0;

    if ( !$ok ) {
        my $errno = $? >> 8;
        local $! = $errno;
        print "

*** The test compile of '$tmp.c' failed: status $?
*** (the status means: errno = $errno or '$!')
*** DO NOT PANIC: this just means that you may get some innocuous
*** compiler warnings.
";
    }
    unlink("$tmp.c");

    }
    return $ok;
}

sub try_cflags {
    my ($ccflags) = @_;
    my $c_prog = "int main () { return 0; }\n";
    print "Checking if $Config{cc} supports \"$ccflags\"...";
    my $result = try_compile($c_prog, ccflags=>$ccflags);
    if ($result) {
        print "yes\n";
        return " $ccflags";
    }
    print "no\n";
    return '';
}

sub check_gsl_version_pkgconfig {

    eval { require PkgConfig };
    if ($@) {
        print "\nI see that you are a CPANTester, you really should install PkgConfig !\n"
            if $ENV{AUTOMATED_TESTING};
        print "\nPkgConfig is currently needed to find GSL for the compilation of this module.\n";
        exit 0;
    }

    modify_env();
    print "\nAsking PkgConfig with ENV{PKG_CONFIG_PATH}=" . ($ENV{PKG_CONFIG_PATH} || '') . "\n\n";
    my $gsl_pkgcfg = PkgConfig->find ('gsl');

    if ($gsl_pkgcfg->errmsg) {
        my $errmsg = $gsl_pkgcfg->errmsg;
        print "
***
*** PkgConfig failed with error message: $errmsg
*** Probably you need to install GSL?
***  On Debian/Ubuntu you can use:
***      sudo apt-get install libgsl0-dev
***  On Mac you can use homebrew (http://brew.sh/):
***      brew install gsl
***  On Fedora/CentOS/RedHat/openSUSE you can use
***      sudo yum install gsl-devel
*** Or get GSL at http://www.gnu.org/software/gsl\n";
        exit 0;
    }

    my $gsl_version = $gsl_pkgcfg->pkg_version;
    if (GSLBuilder::cmp_versions($gsl_version, $MIN_GSL_VERSION) == -1) {
                printf "
***
*** You need to have GSL %s or greater installed. (You have $gsl_version).
*** Get GSL at http://www.gnu.org/software/gsl\n", $MIN_GSL_VERSION;
                exit 0;
            }

    return {
        gsl_version => $gsl_version,
        gsl_prefix  => $gsl_pkgcfg->get_var("prefix"),
        gsl_libs    => join (" ", $gsl_pkgcfg->get_ldflags),
        gsl_cflags  => join (" ", $gsl_pkgcfg->get_cflags),
    };
}

sub check_gsl_version {

    print "Checking for GSL using gsl-config\n";

    my $gsl_version = qx{gsl-config --version};
    if (not defined $gsl_version) {
        print "
***
*** Can't find GSL with gsl-config.
*** Trying with PkgConfig.
    ";
        return undef;
    }

    chomp($gsl_version);
    chomp(my $gsl_prefix  = qx{gsl-config --prefix});
    chomp(my $gsl_cflags  = qx{gsl-config --cflags});
    chomp(my $gsl_libs    = qx{gsl-config --libs});

    my $path_system = catfile(qw/swig system.i/);

    open my $fh, ">", "$path_system" or die "Could not create $path_system : $!";

    my $current_version;

    if ($gsl_version =~ m{\A(\d+(?:\.\d+)+)}) {
        $current_version = $1;
        my @current = split /\./, $current_version;
        print $fh "#define GSL_MAJOR_VERSION $current[0]\n";
        print $fh "#define GSL_MINOR_VERSION $current[1]\n";

        if (GSLBuilder::cmp_versions($current_version, $MIN_GSL_VERSION) == -1) {
            printf "
***
*** You need to have GSL %s or greater installed. (You have $gsl_version).
*** Get GSL at http://www.gnu.org/software/gsl\n", $MIN_GSL_VERSION;
            exit 0;
        } else {
            print "Found GSL $gsl_version (via gsl-config) installed in $gsl_prefix\n";
            return {
                gsl_version => $current_version,
                gsl_prefix  => $gsl_prefix,
                gsl_libs    => $gsl_libs,
                gsl_cflags  => $gsl_cflags,
            };
        }
    } else {
        print "
***
*** Can't parse GSL version $gsl_version.
*** Trying with PkgConfig.
";
        return undef;
    }
    close $fh or die "Could not close $path_system : $!";
}
